drake/solvers/Optimization.h

```diff
@@ -260,13 +447,35 @@ namespace Drake {
       }
     }
 
+    const std::list<std::shared_ptr<Constraint>>& getGenericObjectives() const { return generic_objectives; } // e.g. for snopt_user_fun
+    const std::list<std::shared_ptr<Constraint>>& getGenericConstraints() const { return generic_constraints; } // e.g. for snopt_user_fun
+    std::list<std::shared_ptr<LinearConstraint>> getAllLinearConstraints() const {
+      std::list<std::shared_ptr<LinearConstraint>> conlist = linear_constraints;
+      conlist.insert(conlist.end(),linear_equality_constraints.begin(),linear_equality_constraints.end());
+      return conlist;
+    }
+
   private:
     // note: use std::list instead of std::vector because realloc in std::vector invalidates existing references to the elements
     std::list<DecisionVariable> variables;
+    VariableList variable_views;
+    std::list<std::shared_ptr<Constraint>> generic_objectives;
+    std::list<std::shared_ptr<Constraint>> generic_constraints;
+    std::list<std::shared_ptr<LinearConstraint>> linear_constraints;  // note: does not include linear_equality_constraints
```

Comment:
that's not *quite* right either.  i could have an object of type LinearConstraint which has lb==ub.  But it is not implemented as a derived class LinearEqualityConstraint.